### ISSUE TITLE
Postbox collection resurvey and existence quest

### ISSUE THREAD

**Issue by peternewman** [streetcomplete/StreetComplete#2368]
Interplay between postbox collection resurvey and existence quest.

**How to Reproduce**
Answer an existence quest on an old postbox, I then chose the collection time quest, then pressed dismiss to look at something else. After a brief wait it uploaded did and rescan and the collection times quest had vanished.

However this postbox was over six years old (I've also got resurvey more often on, but I assume it shouldn't inherently matter). Because the existence quest is a higher priority it will override this until it's been resurveyed for the first time I think.

It's mentioned here:
https://github.com/streetcomplete/StreetComplete/blob/ae605ea98dc25f69e47e56646d1ad35bef6a1877/app/src/main/java/de/westnordost/streetcomplete/quests/existence/CheckExistence.kt#L45

Some others say this:
https://github.com/streetcomplete/StreetComplete/blob/ae605ea98dc25f69e47e56646d1ad35bef6a1877/app/src/main/java/de/westnordost/streetcomplete/quests/existence/CheckExistence.kt#L55-L56

FWIW opening hours had changed! Thankfully I took a photo so can fix it manually.

I can't think of a sensible way to fix this apart from only checking postbox existence if collection_times:last_check exists and is sufficiently old, but then you'd never get the existence quest as it would always be beaten by collection_times. But given it can be set to not exist via the collection time isn't it a redundant quest anyway (and seemingly with downsides).

**Versions affected**
SC v28.0-beta1

**Comment by westnordost** [comment]
It's only redundant in countries where the postbox collection quest is asked.

**Comment by peternewman** [comment]
Ah I see, so it's a bit more complex and the mode only needs switching in some places.

Otherwise it will take six years (one existence and one collection resurvey) before they all get checked for changed collection times.

FWIW I had a crash of the collection resurvey quest today, it was a similar workflow of existence first, but I can only assume the background upload took away the quest as I was part way through answering it.

**Comment by westnordost** [comment]
> FWIW I had a crash of the collection resurvey quest today, it was a similar workflow of existence first, but I can only assume the background upload took away the quest as I was part way through answering it.

If you have a stacktrace, post it to me

**Comment by westnordost** [comment]
I will close this as will-not-fix, after all.

Theoretically, the check-existance-quest could be not shown for countries in which the check collection times quests are enabled. Would be a bit hacky, but fine. But: What if the user has no interest in checking collection times but is in a country where that quest is enabled? Then, he'd never be asked. It is not possible to do this "dynamically", depending on whether he currently enabled that or not.

Basically, this is a problem of the OSM API which does not provide for an interface to have timestamps for last-change-dates for individual tags, not just elements. We shouldn't put too much effort into trying to solve this on hte client side.

**Comment by peternewman** [comment]
> If you have a stacktrace, post it to me

No it just did a close and re-open, but it would have been this node https://www.openstreetmap.org/node/457102331 at about 13:30 GMT yesterday if you got anything from Google.

> Theoretically, the check-existance-quest could be not shown for countries in which the check collection times quests are enabled. Would be a bit hacky, but fine. But: What if the user has no interest in checking collection times (=has it disabled) but is in a country where that quest is enabled? Then, he'd never be asked. It is not possible to do this "dynamically", depending on whether he currently enabled that or not.

Fair enough, but in that case I'd argue the correct solution is to just not ask it for those people. One of the other people who also map that area (as some recent research showed) would hopefully have it enabled and handle it.

IMHO the current solution seems to give a worse user experience, quests spontaneously disappearing (and possibly occasionally crashing mid-solve). Otherwise it seems my only other workaround is to prioritise collection times above existence. Plus adding four years to the first resurvey of any old postboxes.

> Basically, this is a problem of the OSM API which does not provide for an interface to have timestamps for last-change-dates for individual tags, not just elements. We shouldn't put too much effort into trying to solve this on hte client side.

I'd probably agree in general, but the behaviour that results with the current operation certainly makes it look like an SC bug.

Alternatively if you don't want to vary the check existence quest, what about only displaying the collection resurvey if the node is either less than 4 years old or has had an existence check done; less optimal, but at least the quest won't spontaneously vanish.

**Comment by RubenKelevra** [comment]
Wondering if this also applies for the vanished shop / opening hour quests 🤔

**Comment by peternewman** [comment]
> Wondering if this also applies for the vanished shop / opening hour quests

I can't see an opening hours overlap between:
https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/quests/existence/CheckExistence.kt#L20-L56
https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/quests/opening_hours/AddOpeningHours.kt#L23-L90

Or:
https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/quests/shop_type/CheckShopType.kt#L13-L21

**Comment by RubenKelevra** [comment]
@peternewman cool! Thanks for checking!

**Comment by peternewman** [comment]
> @peternewman cool! Thanks for checking!

No worries, well done for thinking of it.
